### PR TITLE
fix: improve large payload error handling — better error message and prevent infinite retry and fix conflict with auto chunking

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary
### Problem
When a payload exceeds `MaxPayloadBytes` during externalization, `PayloadInterceptor.MaybeExternalizeAsync` throws an `InvalidOperationException` inside the gRPC `startCallTask`. This exception propagates through `ResponseAsync()` as the **original exception type** (not as `RpcException`), reaching `GrpcDurableTaskWorker.Processor` where it is unhandled. The unhandled exception causes `RunBackgroundTask` to abandon the work item, which gets redelivered by DTS, creating an **infinite retry loop** that never resolves — the orchestration/activity appears permanently stuck.

Additionally, the pre-send `ValidateActionsSize` check blocks any single orchestrator action exceeding 3.9MB, even when large-payload externalization is enabled and the interceptor would handle it. This makes legitimate >3.9MB actions (e.g., 13MB activity output) impossible even with LP configured.

### Changes

**1. `PayloadStorageException` (new class)**
- Introduced `PayloadStorageException : InvalidOperationException` in `Microsoft.DurableTask` namespace for precise exception targeting when payload storage operations fail permanently.
- Thrown by `PayloadInterceptor.MaybeExternalizeAsync` when payload exceeds `MaxPayloadBytes`.
- Thrown by `BlobPayloadStore.DownloadAsync` when the referenced blob is not found (HTTP 404).

**2. Catch `PayloadStorageException` in completion paths**
- **Activity completion** (`CompleteActivityTaskAsync`): Catches `PayloadStorageException` and sends a `FailureDetails` response with `IsNonRetriable = true`, so the orchestration sees a failed activity task instead of getting stuck.
- **Orchestration completion** (`CompleteOrchestratorTaskWithChunkingAsync`): Catches `PayloadStorageException` and completes the orchestration with `OrchestrationStatus.Failed` and a non-retriable `FailureDetails`, preventing infinite retry.
- Exception is matched by fully-qualified type name (`ex.GetType().FullName == "Microsoft.DurableTask.PayloadStorageException"`) because `Worker.Grpc` does not reference `AzureBlobPayloads`.

**3. Skip `ValidateActionsSize` when LP is enabled**
- Added `IsPayloadExternalizationEnabled` boolean property to `GrpcDurableTaskWorkerOptions`.
- Set to `true` by `UseExternalizedPayloads()` in the DI registration.
- When enabled, the pre-send `ValidateActionsSize` check is skipped entirely — the interceptor handles externalization of oversized payloads before the gRPC send.

**4. Fix `TryAddAction` deadlock for oversized first action**
- Changed condition from `currentSize + actionSize > maxChunkBytes` to `currentSize + actionSize > maxChunkBytes && currentSize > 0`.
- Previously, a single action larger than `maxChunkBytes` (e.g., 13MB ScheduleTask with LP) would fail to be added even as the first item in an empty chunk, causing an infinite loop. Now the first action in any chunk is always accepted.

**5. `isChunkedMode` flag for correct chunking behavior**
- When a single oversized action fits in one chunk (via the `TryAddAction` fix), the result has `IsPartial=false` but previously set `ChunkIndex=0`. The DTS backend condition `(isPartial && chunkIndex == 0)` in `PartitionManager.cs` requires chunked responses to have both fields. Setting `ChunkIndex=0` without `IsPartial=true` caused the `ExecutionStartedEvent` to be dropped from the constructed work item.
- Introduced `isChunkedMode` boolean: only set `ChunkIndex` on chunks when `isPartial` has been true at least once (i.e., actual multi-chunk scenario). This matches the existing pattern in the AzureManaged SDK's `AzureManagedOrchestrationService`.

**6. Rename `LargePayloadStorageOptions` properties**
- `ExternalizeThresholdBytes` → `ThresholdBytes`
- `MaxExternalizedPayloadBytes` → `MaxPayloadBytes`
- `CompressPayloads` → `CompressionEnabled`
- Error messages now use generic wording (no config property names) so they work for both Durable Functions (host.json keys) and standalone SDK users.

**7. Fix `MaxDelay` bug in `BlobPayloadStore`**
- Changed `TimeSpan.FromSeconds(MaxDelayMs)` → `TimeSpan.FromMilliseconds(MaxDelayMs)` where `MaxDelayMs = 10_000`.
- The original code resulted in a 10,000 second (2.7 hour) max delay instead of the intended 10 seconds.

**8. Remove dead code in `PayloadInterceptor`**
- Removed the `FailedPrecondition` → `RpcException` mapping in `GetStatus()` that was never reachable (interceptor failures throw through `ResponseAsync()` as original exception types, not via `GetStatus()`).
- Removed corresponding dead `catch (RpcException ... FailedPrecondition)` blocks in the Processor.

## Why is this change needed?

Without this fix, any orchestration or activity whose payload exceeds `MaxPayloadBytes` enters an infinite retry loop with no way to recover — the work item is abandoned and redelivered endlessly, appearing as a permanently stuck instance. This PR ensures such failures are reported as non-retriable task failures, allowing orchestrations to complete with `Failed` status and meaningful error messages.

The `ValidateActionsSize` bypass is needed because with LP enabled, the gRPC interceptor externalizes oversized payloads *before* the gRPC send. The pre-send validation incorrectly rejects these payloads before the interceptor has a chance to process them.

## Issues / work items
- Related to large-payload externalization feature

---

# Project checklist
- [ ] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `release_notes.md`
- [ ] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] Breaking change?
  - [x] If yes:
    - Impact: `LargePayloadStorageOptions` property renames (`ExternalizeThresholdBytes` → `ThresholdBytes`, `MaxExternalizedPayloadBytes` → `MaxPayloadBytes`, `CompressPayloads` → `CompressionEnabled`). These are source-breaking for anyone directly referencing the old property names. Property type `InvalidOperationException` → `PayloadStorageException` for max-size violations (binary-compatible since `PayloadStorageException : InvalidOperationException`).
    - Migration guidance: Update property references to the new names. Exception catches for `InvalidOperationException` from payload operations will still work since `PayloadStorageException` inherits from it.
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot (Claude Opus)
- AI-assisted areas/files: All changed files
- What you changed after AI output: Reviewed all changes, verified exception propagation model through code tracing, validated all 9 E2E scenarios against real DTS

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing

## Automated tests
- **Unit tests** (`LargePayloadStorageOptionsTests.cs`): 6 tests covering defaults, validation bounds, and property behavior for `ThresholdBytes`, `MaxPayloadBytes`, `CompressionEnabled`.
- **Integration tests** (`LargePayloadTests.cs`): 3 new tests:
  - `ActivityOutputExceedsMaxPayload_OrchestrationFailsGracefully` — verifies activity payload-too-large results in `Failed` orchestration, not infinite retry.
  - `OrchestrationOutputExceedsMaxPayload_FailsGracefully` — verifies orchestration payload-too-large results in `Failed` status.
  - `LargeActivityOutput_ExceedsChunkSize_ExternalizesWithLP` — verifies 5MB activity output (exceeding 3.9MB chunk limit) round-trips correctly when LP is enabled.
- Result: All 20+ integration tests pass against in-process sidecar.

## Manual validation (only if runtime/behavior changed)
- Environment: Windows 11, .NET 10.0, Azurite (local blob storage with `--skipApiVersionCheck`), real DTS at `wbtestfilter-gmard3bbapb.eastasia.durabletask.io`
- Steps + observed results:
  1. Started Azurite: `azurite --skipApiVersionCheck`
  2. Ran `LargePayloadConsoleApp` sample with ThresholdBytes=1KB, MaxPayloadBytes=15MB
  3. All 9 scenarios passed against real DTS:

| Scenario | Description | Result |
|----------|-------------|--------|
| 1 | Client oversized input (16MB > 15MB max) | PASS: `PayloadStorageException` thrown immediately |
| 2 | Activity oversized output (16MB > 15MB max) | PASS: Orchestration status=Failed, not stuck |
| 3 | Orchestration oversized output (16MB > 15MB max) | PASS: Orchestration status=Failed, not stuck |
| 4 | 13MB activity I/O (exceeds 3.9MB chunk limit) | PASS: Round-trips correctly, 13631488 bytes |
| 5 | 13MB orchestration input → activity echo | PASS: Input externalized and replayed correctly |
| 6 | 13MB sub-orchestration input → child returns | PASS: Sub-orch action externalized correctly |
| 7 | 13MB external event | PASS: Event externalized and received correctly |
| 8 | 13MB custom status | PASS: Custom status externalized and queryable |
| 9 | 3x 13MB fan-out (real multi-chunk) | PASS: All 3 activities complete, total=40894464 |

---

# Notes for reviewers

### Architecture: Exception propagation in the interceptor
The `PayloadInterceptor` wraps gRPC calls by running `ExternalizeRequestPayloadsAsync` inside `startCallTask` (a `Task.Run`). When externalization fails (e.g., payload too large), the exception propagates through `ResponseAsync()` as the **original exception type** — it is NOT converted to `RpcException`. This means the `GetStatus()` method's error-to-`StatusCode` mapping is dead code and was removed. The Processor must catch the original `PayloadStorageException` type, not `RpcException`.

### Why match by type name instead of direct type reference?
`Worker.Grpc` (where the Processor lives) does not have a project reference to `AzureBlobPayloads` (where `PayloadStorageException` lives). The exception class is in `Microsoft.DurableTask` namespace but physically in the `AzureBlobPayloads` assembly. We match via `ex.GetType().FullName == "Microsoft.DurableTask.PayloadStorageException"` to avoid adding a cross-assembly dependency.

### Why `isChunkedMode`?
The DTS backend (`PartitionManager.cs`) has a condition `(isPartial && chunkIndex == 0)` that expects the first chunk to have both `IsPartial=true` and `ChunkIndex=0`. When a single oversized action fits in one chunk (via `TryAddAction` allowing oversized first items), the response has `IsPartial=false`. If we set `ChunkIndex=0` on this non-partial response, the backend misinterprets it and drops the `ExecutionStartedEvent`, causing the orchestration to start with missing history. The `isChunkedMode` flag ensures `ChunkIndex` is only set when we're actually sending multiple chunks.

### Known follow-up items
1. **Entity completion path**: Entities have a different completion path (`FailureDetails` → sidecar converts to abandon). Entity payload-too-large handling is a separate design discussion.
2. **Broader exception coverage**: Only `PayloadStorageException` is caught. Other exceptions from the interceptor (e.g., transient Azure Storage failures that exhaust retries) would still cause stuck orchestrations. A follow-up could add `ResourceExhausted` catch or a more general safety net.

### CODE REVIEW

File: src/Extensions/AzureBlobPayloads/Options/LargePayloadStorageOptions.cs [hunk 1]
```diff
@@ -85,7 +85,7 @@ public int ExternalizeThresholdBytes
     /// Defaults to 10MB. Requests exceeding this limit will fail fast
     /// with a clear error to prevent unbounded payload growth and excessive storage/network usage.
     /// </summary>
-    public int MaxExternalizedPayloadBytes { get; set; } = 10 * 1024 * 1024;
+    public int MaxPayloadBytes { get; set; } = 10 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Extensions/AzureBlobPayloads/Options/LargePayloadStorageOptions.cs:88]: it is in preview

File: test/Grpc.IntegrationTests/LargePayloadStorageOptionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace Microsoft.DurableTask.Grpc.Tests;
+
+public class LargePayloadStorageOptionsTests
+{
+    [Fact]
+    public void ThresholdBytes_ExactlyOneMiB_DoesNotThrow()
+    {
+        // Arrange
+        LargePayloadStorageOptions options = new();
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply test/Grpc.IntegrationTests/LargePayloadStorageOptionsTests.cs:12]: no noise

File: src/Extensions/AzureBlobPayloads/Interceptors/PayloadInterceptor.cs [hunk 3]
```diff
@@ -174,17 +175,17 @@ public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRe
...
         // Enforce a hard cap to prevent unbounded payload sizes
-        if (size > this.options.MaxExternalizedPayloadBytes)
+        if (size > this.options.MaxPayloadBytes)
         {
-            throw new InvalidOperationException(
-                $"Payload size {size / 1024} kb exceeds the configured maximum of {this.options.MaxExternalizedPayloadBytes / 1024} kb. " +
-                "Consider reducing the payload or increase MaxExternalizedPayloadBytes setting.");
+            throw new PayloadStorageException(
+                $"Payload size {size / 1024} KB exceeds the configured maximum of {this.options.MaxPayloadBytes / 1024} KB. " +
+                "Reduce the payload size or increase the max payload size limit.");
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Extensions/AzureBlobPayloads/Interceptors/PayloadInterceptor.cs:187]: thats intentional, customer should be referring to the doc themselves

File: src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs [hunk 4]
```diff
@@ -853,7 +853,32 @@ await this.client.AbandonTaskActivityWorkItemAsync(
...
+                    InstanceId = instance.InstanceId,
+                    TaskId = request.TaskId,
+                    FailureDetails = new P.TaskFailureDetails
+                    {
+                        ErrorType = typeof(InvalidOperationException).FullName,
+                        ErrorMessage = ex.Message,
+                        StackTrace = ex.ToString(),
+                        IsNonRetriable = true,
+                    },
+                    CompletionToken = completionToken,
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs]: no noise

**Comment by YunchuWang** [comment]
## E2E Validation Results (2026-04-01, real DTS: wbtestfilter-gmard3bbapb.eastasia.durabletask.io)

**Config:** ThresholdBytes=1KB, MaxPayloadBytes=15MB, Azurite (UseDevelopmentStorage=true)

| # | Scenario | Validates | Status | Output |
|---|---|---|---|---|
| 1 | Client 16MB input | PayloadStorageException thrown immediately | :white_check_mark: PASS | PayloadStorageException: Payload size 16384 KB exceeds the configured maximum of 15360 KB |
| 2 | Activity 16MB output | Orchestration fails gracefully (not infinite retry) | :white_check_mark: PASS | Status: Failed / Error: Task ProduceOversized failed...Payload size 16384 KB exceeds... |
| 3 | Orchestration 16MB output | Orchestration fails gracefully (not infinite retry) | :white_check_mark: PASS | Status: Failed / Error: Payload size 16384 KB exceeds... |
| 4 | 13MB activity input->echo->orch output | ValidateActionsSize bypass + isChunkedMode (single oversized as non-chunked) + ScheduleTask/ActivityResponse/CompleteOrchestration externalization | :white_check_mark: PASS | Status: Completed / Output length: 13631488 |
| 5 | 13MB orchestration input->activity echo | CreateInstanceRequest.Input externalization + replay with token in history | :white_check_mark: PASS | Status: Completed |
| 6 | 13MB sub-orchestration input | CreateSubOrchestration.Input externalization (>3.9MB action) | :white_check_mark: PASS | Status: Completed |
| 7 | 13MB external event | RaiseEventRequest.Input externalization + WaitForExternalEvent resolution | :white_check_mark: PASS | Status: Completed |
| 8 | 13MB custom status | OrchestratorResponse.CustomStatus externalization + GetInstance resolution | :white_check_mark: PASS | Status: Completed |
| 9 | 3x 13MB activity inputs (multi-chunk) | Real chunked LP completion (isChunkedMode=true, ChunkIndex=0,1,...) | :white_check_mark: PASS | Status: Completed / total=40894464 |

**Integration tests:** 20/20 passed (6 unit + 14 integration with in-process sidecar)

image

File: samples/LargePayloadConsoleApp/Program.cs [hunk 5]
```diff
@@ -94,10 +96,36 @@
...
+
+        // Overflow: orchestration output > MaxPayloadBytes
+
+        tasks.AddOrchestratorFunc<object?, string>("OrchestrationProducesOversized", (ctx, _) =>
+
+        {
+
+            return Task.FromResult(new string('P', 60 * 1024));
+
+        });
```
[the remark this answers is not in the document]
Comment (cgillum) [reply samples/LargePayloadConsoleApp/Program.cs]: @YunchuWang was this resolved by mistake? I'm still seeing lots of extra whitespace, which is a distraction.

File: src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs [hunk 6]
```diff
@@ -853,7 +853,32 @@ await this.client.AbandonTaskActivityWorkItemAsync(
             // Stop the trace activity here to avoid including the completion time in the latency calculation
             traceActivity?.Stop();
 
-            await this.client.CompleteActivityTaskAsync(response, cancellationToken: cancellation);
+            try
+            {
+                await this.client.CompleteActivityTaskAsync(response, cancellationToken: cancellation);
+            }
+            catch (InvalidOperationException ex) when (ex.GetType().FullName == "Microsoft.DurableTask.PayloadStorageException")
```
Comment (cgillum) [inline src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs]: Hmm...this looks like a leaky design issue to me and I don't think it's acceptable. It would be better if the interceptor simply replaced the response with a failure payload directly rather than throwing an exception. That, or we need to reconsider the design of large payload handling.

File: src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs [hunk 7]
```diff
@@ -1002,57 +1029,106 @@ static bool TryAddAction(
...
+                        chunkedResponse.OrchestrationTraceContext = response.OrchestrationTraceContext;
+                    }
+
+                    chunkIndex++;
+
+                    // Send the chunk
+                    await this.client.CompleteOrchestratorTaskAsync(chunkedResponse, cancellationToken: cancellationToken);
                 }
+            }
+            catch (InvalidOperationException ex) when (ex.GetType().FullName == "Microsoft.DurableTask.PayloadStorageException")
```
Comment (cgillum) [inline src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs]: Same feedback as above.

File: src/Extensions/AzureBlobPayloads/DependencyInjection/DurableTaskWorkerBuilderExtensions.AzureBlobPayloads.cs [hunk 8]
```diff
@@ -80,6 +80,7 @@ static IDurableTaskWorkerBuilder UseExternalizedPayloadsCore(IDurableTaskWorkerB
                 }
 
                 opt.Capabilities.Add(P.WorkerCapability.LargePayloads);
+                opt.IsPayloadExternalizationEnabled = true;
```
Comment (cgillum) [inline src/Extensions/AzureBlobPayloads/DependencyInjection/DurableTaskWorkerBuilderExtensions.AzureBlobPayloads.cs]: Why do we need this new property when we already have the Capabilities to tell us that we support large payloads?

**Review by cgillum — CHANGES_REQUESTED** [review]

File: src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs (hunk 6 above)
Comment (YunchuWang) [reply src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs]: Thanks for the feedback Chris @cgillum . Agreed — the Processor shouldn't need to know about payload storage internals.

I've refactored this in commit 4213b0d. The interceptor (\AzureBlobPayloadsSideCarInterceptor\) now handles permanent failures internally by replacing the request with a failure payload:

- **\ActivityResponse\**: On \PayloadStorageException\, clears \Result\ and sets \FailureDetails\ (with \IsNonRetriable = true\) directly on the response. The gRPC call proceeds normally with the failure payload — no exception escapes.
- **\OrchestratorResponse\**: On \PayloadStorageException\, clears all \Actions\, resets \IsPartial\/\ChunkIndex\, and replaces with a single \CompleteOrchestration(Failed)\ action containing the failure details.

Both catch blocks in the Processor are removed — it no longer has any awareness of payload storage exceptions.

Transient failures (\RequestFailedException\ from Azure SDK after retry exhaustion, \TimeoutException\, etc.) are intentionally **not** caught by the interceptor — they propagate naturally and cause the work item to be abandoned and redelivered, which is the correct behavior for infrastructure issues that may self-resolve.

For the \IsPayloadExternalizationEnabled\ question — also addressed in the same commit. Replaced it with \Capabilities.Contains(WorkerCapability.LargePayloads)\, removed the extra property entirely.

All 9 E2E scenarios verified against real DTS after this change.

File: src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs [hunk 9]
```diff
@@ -360,4 +407,28 @@ async Task ResolveEventPayloadsAsync(P.HistoryEvent e, CancellationToken cancell
...
+        }
+
+        // Azure SDK does not retry 4xx errors (except 408/429). If we receive a 4xx
+        // RequestFailedException, it means the request is fundamentally invalid
+        // (e.g., 401 bad credentials, 403 missing RBAC role, 404 account/container not found).
+        // These will never succeed on retry with the same configuration.
+        if (ex is RequestFailedException rfe && rfe.Status >= 400 && rfe.Status < 500)
+        {
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs]: I agree with this feedback. But do we have a mechanism to do backoff/retry or have a max number of retry attempts?

File: samples/LargePayloadConsoleApp/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (YunchuWang) [reply samples/LargePayloadConsoleApp/Program.cs]: yes it is cleaned up now

File: src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs [hunk 10]
```diff
@@ -957,7 +957,9 @@ async Task CompleteOrchestratorTaskWithChunkingAsync(
                 return null;
             }
 
-            P.TaskFailureDetails? validationFailure = ValidateActionsSize(response.Actions, maxChunkBytes);
+            P.TaskFailureDetails? validationFailure = this.worker.grpcOptions.Capabilities.Contains(P.WorkerCapability.LargePayloads)
+                ? null
+                : ValidateActionsSize(response.Actions, maxChunkBytes);
             if (validationFailure != null)
             {
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Grpc/GrpcDurableTaskWorker.Processor.cs:964]: you are over defensive.

File: test/Grpc.IntegrationTests/LargePayloadStorageOptionsTests.cs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply test/Grpc.IntegrationTests/LargePayloadStorageOptionsTests.cs:13]: no bro no need to do that

File: src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs [hunk 12]
```diff
@@ -39,10 +40,56 @@ protected override async Task ExternalizeRequestPayloadsAsync<TRequest>(TRequest
...
+                        CompleteOrchestration = new P.CompleteOrchestrationAction
+                        {
+                            OrchestrationStatus = P.OrchestrationStatus.Failed,
+                            FailureDetails = new P.TaskFailureDetails
+                            {
+                                ErrorType = ex.GetType().FullName,
+                                ErrorMessage = ex.Message,
+                                StackTrace = ex.ToString(),
+                                IsNonRetriable = true,
+                            },
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs:88]: done

File: src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs [hunk 13]
```diff
@@ -39,10 +40,56 @@ protected override async Task ExternalizeRequestPayloadsAsync<TRequest>(TRequest
...
+                    // Replace with a failure response so the orchestration sees a failed activity
+                    // instead of the work item being abandoned and redelivered indefinitely.
+                    r.Result = null;
+                    r.FailureDetails = new P.TaskFailureDetails
+                    {
+                        ErrorType = ex.GetType().FullName,
+                        ErrorMessage = ex.Message,
+                        StackTrace = ex.ToString(),
+                        IsNonRetriable = true,
+                    };
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs:59]: done

File: src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Extensions/AzureBlobPayloads/Interceptors/AzureBlobPayloadsSideCarInterceptor.cs]: as discussed offline, yes we have blob client built in retries configured. surfacing up these 408/429 for retries if current round of sdk retries exhausted. will add a followup workitem to do poison queue for messages retried too many times.
